### ISSUE TITLE
Strange growth in the loss

### ISSUE THREAD

**Issue by Zeta36** [ibab/tensorflow-wavenet#143]
This is the case:

I start to train a little dataset so I can get a very small (overfitted) loss. In the beginning all goes well. I start with a ~5.5 error cost, and while training the loss drop gradually until I reach a near ~0.05 loss. but, and this is the question: if I continue training after this low loss has been reached; after some hundred steps more, the cost begins to **grow** instead of staying near 0!! (sometimes, the loss go up to a value much above 5??)

I don't know what can cause this strange behavior, but I guess it's not a normal thing.

What do you think?

**Comment by Zeta36** [comment]
May be the case that simply the Reader get out of data to feed into the sample_placeholder (as I use a little data set)?

**Comment by ibab** [comment]
@Zeta36: From what I've seen, the problem is that the ReLU layers at the end cut off the entire network output. The following things seem to help:
 - use a smaller learning rate
 - use regularization
 - replace the ReLU layers with something that's nonzero almost everywhere like Tanh

**Comment by jyegerlehner** [comment]
> From what I've seen, the problem is that the ReLU layers at the end cut off the entire network output.

If that's what's happening, I wonder if an option to make the "post-processing" ReLU layers res blocks would help. Residual blocks usually make things better anyway, allowing activations and gradients to go around the layer.

**Comment by jyegerlehner** [comment]
@Zeta36 @ibab 

**TL;DR:** I think I may have a fix for what Zeta36 describes. They are in [PR 128](https://github.com/ibab/tensorflow-wavenet/pull/128) for rmsprop and [PR 147](https://github.com/ibab/tensorflow-wavenet/pull/147) for adam.

### Explanation

I've seen the behavior Zeta36 is describing in our test/test_model.py. When that test uses adam or rmsprop, I would see the loss drop and drop till a small number, then jump up to a large loss at some random time. 

You can reproduce that problematic behavior if you change (what at the time of this writing in master) [MomentumOptimizer  to AdamOptimizer](https://github.com/ibab/tensorflow-wavenet/blob/master/test/test_model.py#L58), and make the learning rate 0.002. Uncomment the statements that [print loss](https://github.com/ibab/tensorflow-wavenet/blob/master/test/test_model.py#L73).

If you run the test with 
```
python test/test_model.py
```
every second or third time or so that you run the test you will see the loss will drop and then at some point jump up to a larger value, and sometimes cause the test to fail. I "worked around" that problem by futzing with the learn rate and number of training iterations we run in the test until it would reliably pass.

Anyway, I think I've found the cause. If you look at the tensorflow implementations in tensorflow of [rmsprop](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/rmsprop.py#L26) and [adam](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/adam.py#L60) you will see they compute the change to a weight by dividing by a some sinister lag-filtered rate-of-change or error magnitude.  When the error or rate of change of error gets small, or even zero, near the bottom of the error basin, then the denominator gets close to zero. The only thing saving us from a NaN or Inf is they add an epsilon. That epsilon defaults to 1e-10 for rmsprop and 1e-8 for adam. That's enough to make the change to our parameter a big number, presumably big enough give us a large loss. 

So in [PR 128](https://github.com/ibab/tensorflow-wavenet/pull/128) I specified a larger epsilon for rmsprop, and in [PR 147](https://github.com/ibab/tensorflow-wavenet/pull/147) for adam. I found that these changes fix the problem of randomly increasing loss during the tests in test_model.py.

**Comment by ibab** [comment]
This should be fixed by #128 and #147 now :)
